### PR TITLE
Enables URL parsing of p5.js version and add-on libraries

### LINKED ISSUE #3647: Enable URL parsing of p5.js version
### Increasing Access

Currently, educators who want to use p5.js 2.0 have to instruct students to always open library management settings, and select the 2.0 version. URL parsing on the editor would enable educational materials to list the URL with the version already specified. Currently, this is especially relevant for using 2.x, but in the future, it would be important for any non-default version: Once 2.x is default, for example, a tutorial that requires 1.x would also benefit from this feature.

### Feature request details

Parse URL, so that `editor.p5js.org?version=2.0.4` opens a sketch with that library set. If no version, or a nonexistent version, is provided, then the editor should open a blank sketch with the current default version - as it does now. In addition to implementing this functionality, make sure that as much of this feature as possible is covered with [automated test(s)](https://github.com/processing/p5.js-web-editor/blob/develop/contributor_docs/testing.md).

Currently, version is only controlled as follows:

Image

With URL parsing, it would still be managed in this dialog, but also through URL.

Potential future extensions - these are open to discussion, and any contributors working on the main issue are welcome to share their thoughts:
* interpret not only specific version numbers (as above), but also latest patch `editor.p5js.org?version=2.0` or minor version (`editor.p5js.org?version=2` )
* include add-on libraries; for example, p5.sound.js, or the three 1.x compatibility add-ons provide 1.x-style functionality in a 2.x sketch, which may also be needed in some cases.

### PR DESCRIPTION
**Pull request by Hugo-RM**
Fixes #3647

## Changes

- Adds URL parser to allow p5.js version to be configured from URL.
- Includes URL parser test file to test edge cases and fallback response.

### Potential future extensions addressed

- Supports latest patch and minor versions as requested in issue #3647
- URL parser allows for toggling on and off add-on libraries with robust variants allowed.

## Demo Video

[Demo Video](https://youtu.be/Kg6xKqNQFTQ) (Wasn't able to upload it directly sorry!)

Video demonstrates:
- Default behavior (remains the same)
- Different version inputs
- Add-on toggling functionality
- Fallback behavior for invalid inputs
- Edge case handling for the sound add-on URL with p5 version 2.x vs 1.x/0.x

## Note about TypeScript migration

We began working on this before contributions to the migration effort were open to the public (October 31, 2025). As such, our implementation is in JavaScript. Given that it is now past October 31, we can:

- Edit our changes to utilize TypeScript
- Open a separate issue for migrating the URL parser to TypeScript
- Leave the changes as is
- Another approach (your choice)

Let us know your preference.

## I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] has no test errors (`npm run test`)
* [x] is from a uniquely-named feature branch and is up to date with the  `develop` branch.
* [x] is descriptively named and links to an issue number, i.e. `Fixes #123`
* [x] meets the standards outlined in the [accessibility guidelines](https://github.com/processing/p5.js-web-editor/blob/develop/contributor_docs/accessibility.md)

### CODE REVIEW

File: server/domain-objects/createDefaultFiles.js [hunk 1]
```diff
@@ -8,11 +8,35 @@ function draw() {
   background(220);
 }`;
 
-export const defaultHTML = `<!DOCTYPE html>
+export function defaultHTML({
```
Comment (raclim) [inline server/domain-objects/createDefaultFiles.js:21]: This is a great change—it adds much more flexibility in creating the default files and the changes here are easy to follow!

**Comment by raclim** [comment]
Thanks so much for your work and patience @Hugo-RM and @Enochteo! I'm sorry about the delay in getting back on this! 

I really appreciate the thorough documentation and clarity of your process—I feel that it conveys a clear understanding of the issue, potential edge cases, and where it fits in relation to other ongoing work. I think the proposed solution overall works really well by adding in the URL parsing functionality in a simple and consistent way, and is most likely good to go! 

Regarding refactoring it to TypeScript, I think going with the second option of "opening a separate issue for migrating the URL parser to TypeScript" would be best! Please feel free to go ahead with opening a new issue for it if that would be something the both of you would be interested in! 

Tagging @ksen0 in case you have any other thoughts here—also linking [a URL to a Release environment for this PR](https://app-tedcacb.release.editor.p5js.org/) to quickly test it out!

**Review by raclim — DISMISSED** [review]

**Comment by ksen0** [comment]
Thanks so much for this thorough and well=documented work @Hugo-RM and @Enochteo! I also really appreciated the video.

I noticed that the addon flags (I tried to follow how the video presented it) did not work for me:

* https://app-tedcacb.release.editor.p5js.org/?version=2.1.1&data=ON
* https://app-tedcacb.release.editor.p5js.org/?version=2.1.1&data=on
* https://app-tedcacb.release.editor.p5js.org/?version=2.1.1&data=1

all still have the default off on data for me. If I've misunderstood the usage, how should it look like instead?
In any case, really great feature, excited to see it live!

**Comment by Hugo-RM** [comment]
Thank you for the kind words!

@ksen0 I apologize for the issue, you were using it correctly. There is an issue I missed.

### Issue

After looking through the recently merged commits, I realized that the issue with the data & events add-on stems from the recent change from version 0.1.2 to 0.2.0 in their URLs. This means preload and shapes have also stopped working.

#### `server/domain-objects/createDefaultFiles.js`
```.js
    preload                                               // Should be 0.2.0 now
      ? `<script src="https://cdn.jsdelivr.net/npm/p5.js-compatibility@0.1.2/src/preload.js"></script>`
      : '',
    shapes                                                // Should be 0.2.0 now
      ? `<script src="https://cdn.jsdelivr.net/npm/p5.js-compatibility@0.1.2/src/shapes.js"></script>`
      : '',
    data                                                  // Should be 0.2.0 now
      ? `<script src="https://cdn.jsdelivr.net/npm/p5.js-compatibility@0.1.2/src/data.js"></script>`
```

### Testing

I tested this by simply switching the version in `server/domain-objects/createDefaultFiles.js`. URLs with the outdated versions would not work but those with updated ones would. 

### Proposed Fix

I see two possible fixes but would like to ask which would be preferred.

- Move the URLs to `common/` and export it from there.
  - I imagine this as similar to how `currentP5Version` is implemented and used currently.
  - Changes apply to all usages of the variables.
- Update the URL strings in `createDefaultFiles.js` to use 0.2.0 (temporary fix, not ideal imo).

I believe importing the URLs from `useP5Version.jsx` into `createDefaultFiles.js` would not be possible because a server file is importing from a client file. If this is not an issue, I can try doing this instead if preferred.

**Comment by raclim** [comment]
I think moving forward with the first proposed solution "Move the URLs to `common/` and export it from there." would be best! Please feel free to go ahead with updating the PR and notifying us when it's done.

**Comment by Hugo-RM** [comment]
@raclim @ksen0 I updated the PR to have the URL constants moved to a new file `common/p5URLs.js`. The add-on toggle issue should now be fixed. Please let me know if there are any issues!

**Comment by ksen0** [comment]
Great work @Hugo-RM thanks! I've tested this and both version and add-on settings worked great.

**Review by ksen0 — APPROVED** [review]
